gatsby-node.js

```diff
@@ -171,83 +130,86 @@ async function doQuery({
     enablePartialUpdates,
   });
 
-  /* Use to keep track of what to remove afterwards */
-  const toRemove = {};
+  let toIndex = {}; // used to track objects that should be added / updated
+  const toRemove = {}; // used to track objects that are stale and should be removed
 
-  setQueryStatus('Executing query...');
-  const result = await graphql(query);
-  if (result.errors) {
-    report.panic(
-      `failed to index to Algolia, errors:\n ${JSON.stringify(result.errors)}`,
-      result.errors
-    );
-  }
+  if (enablePartialUpdates !== true) {
+    // enablePartialUpdates isn't true, so index all objects
+    toIndex = { ...allObjectsMap };
+  } else {
+    // iterate over each query to determine which data are fresh
+    setStatus(activity, `Starting Partial updates...`);
 
-  const objects = (await transformer(result)).map(object => ({
-    objectID: object.objectID || object.id,
-    ...object,
-  }));
+    // get all match fields for all queries to minimize calls to the api
+    const allMatchFields = getAllMatchFields(queries, mainMatchFields);
 
-  if (objects.length > 0 && !objects[0].objectID) {
-    report.panic(
-      `failed to index to Algolia. Query results do not have 'objectID' or 'id' key`
+    // get all indexed objects matching all matched fields
+    const indexedObjects = await fetchAlgoliaObjects(
+      indexToUse,
+      allMatchFields
     );
-  }
-
-  setQueryStatus(`graphql resulted in ${objects.length} records`);
-
-  let hasChanged = objects;
-  if (enablePartialUpdates) {
-    setQueryStatus(`Starting Partial updates...`);
-
-    const algoliaObjects = await fetchAlgoliaObjects(indexToUse, matchFields);
 
-    const nbMatchedRecords = Object.keys(algoliaObjects).length;
-    setQueryStatus(`Found ${nbMatchedRecords} existing records`);
-
-    if (nbMatchedRecords) {
-      hasChanged = objects.filter(curObj => {
-        if (matchFields.every(field => Boolean(curObj[field]) === false)) {
-          report.panic(
-            'when enablePartialUpdates is true, the objects must have at least one of the match fields. Current object:\n' +
-              JSON.stringify(curObj, null, 2) +
-              '\n' +
-              'expected one of these fields:\n' +
-              matchFields.join('\n')
-          );
+    // iterate over each query
+    for (const [i, { matchFields = mainMatchFields }] of queries.entries()) {
```

Comment:
is queries here the results or the requests? (sorry, I'm confused :p)

Reply:
Sorry for the slow response, I've been on vacation. `queries` here are the requests initially passed in to the plugin, and `objectMapsByQuery` is an array of the same length, where the `n`th index matches the corresponding query, so `objectMapsByQuery[i]` are the results of running `queries[i]`. If there's a more intuitive way to organize that, let me know!